File: src/services/compareOptions.ts
import { OptionQuote } from "../exchanges";

export type OptionMap = Map<string, OptionQuote[]>;

export type OptionSpread = {
  symbol: string;
  symbol_date: string;
  instrument: string;
  buy_from_instrument: string;
  sell_to_instrument: string;
  buy_from_underlying_price: number;
  sell_to_underlying_price: number;
  strike: number;
  bid_raw: number;
  ask_raw: number;
  bid_price: number;
  ask_price: number;
  type: string;
  expiration: string;
  buy_from: string;
  sell_to: string;
  buy_ask: number;
  sell_bid: number;
  spread: number;
  apr: number;
  maxSize: number;
  buyLink?: string;
  sellLink?: string;
};

function isValidQuote(q: OptionQuote): q is Required<OptionQuote> {
  return (
    q.bid_price !== undefined &&
    q.ask_price !== undefined &&
    q.bid_qty !== undefined &&
    q.ask_qty !== undefined &&
    q.bid_price > 0 &&
    q.ask_price > 0
  );
}

export function groupOptionsByInstrument(allQuotes: OptionQuote[]): OptionMap {
  const map: OptionMap = new Map();

  for (const quote of allQuotes) {
    const key = quote.normalized_name;
    if (!map.has(key)) map.set(key, []);
    map.get(key)!.push(quote);
  }

  return map;
}

export function compareOptions(optionGroups: OptionQuote[][]): OptionSpread[] {
  const now = Date.now();

  const results = optionGroups
    .map((quotes) => {
      if (quotes.length < 2) return null; // need at least 2 exchanges to compare

      // Ensure all are same instrument
      const nameSet = new Set(quotes.map((q) => q.normalized_name));
      if (nameSet.size > 1) {
        throw new Error(
          `Mismatched instruments in group: ${[...nameSet].join(", ")}`
        );
      }

      const instrument = quotes[0].normalized_name;
      const expiry = quotes[0].expiry;
      const daysToExp = (expiry - now) / (1000 * 60 * 60 * 24);

      // Filter only valid quotes (have prices and qty)
      const valid = quotes.filter(isValidQuote);

      if (valid.length < 2) return null;

      // Find lowest ask (best buy) and highest bid (best sell)
      const lowestAsk = valid.reduce((min, q) =>
        q.ask_price < min.ask_price ? q : min
      );
      const highestBid = valid.reduce((max, q) =>
        q.bid_price > max.bid_price ? q : max
      );

      if (lowestAsk.exchange === highestBid.exchange) return null; // no cross-exchange arb

      const priceDiffPct =
        ((highestBid.bid_price - lowestAsk.ask_price) / lowestAsk.ask_price) *
        100;
      const netDiffPct =
        priceDiffPct -
        (lowestAsk.taker_fee_rate + highestBid.taker_fee_rate) * 100;

      if (netDiffPct <= 0) return null; // no profit

      const apr = (netDiffPct / daysToExp) * 365;

      // Compute max executable notional size (based on liquidity)
      const maxSize = Math.min(
        lowestAsk.ask_qty * lowestAsk.ask_price,
        highestBid.bid_qty * highestBid.bid_price
      );

      const name = quotes[0].normalized_name;
      const firstDash = name.indexOf("-");
      const secondDash =
        firstDash === -1 ? -1 : name.indexOf("-", firstDash + 1);
      const symbol_date = secondDash === -1 ? name : name.slice(0, secondDash);

      return {
        symbol: name.slice(0, 3),
        symbol_date,
        instrument,
        buy_from_instrument: lowestAsk.instrument_name,
        sell_to_instrument: highestBid.instrument_name,
        buy_from_underlying_price: +lowestAsk.underlying_price,
        sell_to_underlying_price: +highestBid.underlying_price,
        strike: quotes[0].strike,
        bid_raw: highestBid.bid_price_raw,
        ask_raw: lowestAsk.ask_price_raw,
        type: quotes[0].option_type,
        expiration: `${daysToExp.toFixed(1)}d`,
        buy_from: lowestAsk.exchange,
        sell_to: highestBid.exchange,
        buy_ask: +lowestAsk.ask_price,
        sell_bid: +highestBid.bid_price,
        spread: +netDiffPct.toFixed(2),
        apr: +apr.toFixed(2),
        maxSize: +maxSize.toFixed(2),
      };
    })
    .filter(Boolean);

  return results as OptionSpread[];
}

File: src_python/services/compare_options.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from time import time

from src_python.exchanges import OptionQuote


@dataclass(slots=True)
class OptionSpread:
    symbol: str
    symbol_date: str
    instrument: str
    buy_from_instrument: str
    sell_to_instrument: str
    buy_from_underlying_price: float
    sell_to_underlying_price: float
    strike: float
    bid_raw: float | None
    ask_raw: float | None
    bid_price: float
    ask_price: float
    type: str
    expiration: str
    buy_from: str
    sell_to: str
    buy_ask: float
    sell_bid: float
    spread: float
    apr: float
    maxSize: float
    buyLink: str | None = None
    sellLink: str | None = None


OptionMap = dict[str, list[OptionQuote]]


def _is_valid_quote(q: OptionQuote) -> bool:
    return (
        q.bid_price is not None
        and q.ask_price is not None
        and q.bid_qty is not None
        and q.ask_qty is not None
        and q.bid_price > 0
        and q.ask_price > 0
    )


def group_options_by_instrument(all_quotes: list[OptionQuote]) -> OptionMap:
    grouped: defaultdict[str, list[OptionQuote]] = defaultdict(list)
    for quote in all_quotes:
        grouped[quote.normalized_name].append(quote)
    return dict(grouped)


def compare_options(option_groups: list[list[OptionQuote]]) -> list[OptionSpread]:
    now_ms = int(time() * 1000)
    results: list[OptionSpread] = []

    for quotes in option_groups:
        if len(quotes) < 2:
            continue

        names = {q.normalized_name for q in quotes}
        if len(names) > 1:
            raise ValueError(f"Mismatched instruments in group: {', '.join(sorted(names))}")

        name = quotes[0].normalized_name
        expiry = quotes[0].expiry
        days_to_exp = (expiry - now_ms) / (1000 * 60 * 60 * 24)
        if days_to_exp <= 0:
            continue

        valid = [q for q in quotes if _is_valid_quote(q)]
        if len(valid) < 2:
            continue

        lowest_ask = min(valid, key=lambda q: q.ask_price or float("inf"))
        highest_bid = max(valid, key=lambda q: q.bid_price or 0)

        if lowest_ask.exchange == highest_bid.exchange:
            continue

        assert lowest_ask.ask_price is not None
        assert highest_bid.bid_price is not None
        assert lowest_ask.ask_qty is not None
        assert highest_bid.bid_qty is not None

        price_diff_pct = ((highest_bid.bid_price - lowest_ask.ask_price) / lowest_ask.ask_price) * 100
        net_diff_pct = price_diff_pct - (lowest_ask.taker_fee_rate + highest_bid.taker_fee_rate) * 100
        if net_diff_pct <= 0:
            continue

        apr = (net_diff_pct / days_to_exp) * 365
        max_size = min(
            lowest_ask.ask_qty * lowest_ask.ask_price,
            highest_bid.bid_qty * highest_bid.bid_price,
        )

        first_dash = name.find("-")
        second_dash = name.find("-", first_dash + 1) if first_dash != -1 else -1
        symbol_date = name if second_dash == -1 else name[:second_dash]

        results.append(
            OptionSpread(
                symbol=name[:3],
                symbol_date=symbol_date,
                instrument=name,
                buy_from_instrument=lowest_ask.instrument_name,
                sell_to_instrument=highest_bid.instrument_name,
                buy_from_underlying_price=float(lowest_ask.underlying_price or 0),
                sell_to_underlying_price=float(highest_bid.underlying_price or 0),
                strike=quotes[0].strike,
                bid_raw=highest_bid.bid_price_raw,
                ask_raw=lowest_ask.ask_price_raw,
                bid_price=float(highest_bid.bid_price),
                ask_price=float(lowest_ask.ask_price),
                type=quotes[0].option_type,
                expiration=f"{days_to_exp:.1f}d",
                buy_from=lowest_ask.exchange,
                sell_to=highest_bid.exchange,
                buy_ask=float(lowest_ask.ask_price),
                sell_bid=float(highest_bid.bid_price),
                spread=round(net_diff_pct, 2),
                apr=round(apr, 2),
                maxSize=round(max_size, 2),
            )
        )

    return results

